man/skel.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "SKEL" "1" "July 2018" "" ""
.
.SH "SYNOPSIS"
skel [\-h] [\-o OPENER] [\-c CLOSER] [\-d FILE] [\-p PATH] [\-e] [\-x] [TEMPLATE] [OUT]
.
.SH "DESCRIPTION"
skel reads a template from stdin or a file given on the command line, replacing expansion patterns with the contents of variables from the environment\.
.
.P
For example,
.
.IP "" 4
.
.nf

$ skel TEMPLATE_FILE
.
.fi
.
.IP "" 0
.
.P
will cat TEMPLATE_FILE, replacing instances of \fB#{HOME}\fR with the contents of \fB#{HOME}\fR in the environment\.
.
.P
If TEMPLATE_FILE is a directory, then skel will run \fBls \-1\fR on the directory (as a convenience) and exit\.
.
.SH "OPTIONS"
.
.TP
\fB\-h\fR
Print a usage summary and exit\.
.
.TP
\fB\-o OPENER\fR
Replace string used to mark the start of expansion patterns\. The opener defaults to \'#{\', and cannot contain \':\'\.
.
.TP
\fB\-c CLOSER\fR
Replace string used to mark the end of expansion patterns\. The closer defaults to \'}\', and cannot contain \':\'\.
.
.TP
\fB\-d\fR
A file which should be used as defaults\. See DEFAULT FILE below\.
.
.TP
\fB\-p\fR
Path to the skeletons\' closet\.
.
.TP
\fB\-e\fR
Exit with an error if variables are undefined; otherwise, undefined variables are output as \'#{VARNAME}\'\.
.
.TP
\fB\-x\fR
Enable execution patterns\. See ATTRIBUTES below\.
.
.SH "EXIT STATUS"
Returns 0 when the template has been output\.
.
.P
Returns 1 if \fB\-e\fR was used and a variable was undefined, a variable name is too long, or a default file could not be read, or the output file could not be opened\.
.
.P
Returns 2 if \fB\-x\fR was used and a command failed to execute\.
.
.SH "ENVIRONMENT"
Since variables are expanded from the shell environment, it\'s easy to pass substitutions to skel using env(1):
.
.IP "" 4
.
.nf

$ env FOO="foo" skel TEMPLATE_FILE
.
.fi
.
.IP "" 0
.
.SH "FILES"
Unless the \fB\-p\fR option specifies a path, skel will check for the template file in the following paths, if present:
.
.IP "" 4
.
.nf

\&\.
${SKEL_CLOSET} (default: ~/\.dem_bones)
${SKEL_SYSTEM_CLOSET} (default: /usr/local/share/skel/)
.
.fi
.
.IP "" 0
.
.SH "NOTES"
Expansion patterns are of the form \fB#{:ATTRIBUTES:VARNAME:DEFAULT}\fR, where the \fB:ATTRIBUTES:\fR and \fB:DEFAULT\fR sections can be omitted\. The expansion will be replaced with the value of \fBVARNAME\fR in the shell environment, as modified by the attributes\. If the variable is undefined, then the default will be used (if given), otherwise the expansion will be printed as "#{VARNAME}"\.
.
.P
Variable names are limited to alphanumeric characters and \fB_\fR\. The expansion opener and closer strings can be changed with \fB\-o\fR and \fB\-c\fR (they default to \fB#{\fR and \fB}\fR), but cannot contain \fB:\fR\.
.
.SS "ATTRIBUTES"
.
.nf

l: convert to lowercase
u: convert to uppercase
n: eliminate trailing newline
x: Pass expansion body to shell and insert output
.
.fi
.
.P
The \'x\' attribute requires \fBskel \-x\fR to avoid accidentally running dangerous commands; it\'s disabled by default\.
.
.P
For example, \fB#{:xn:date +%Y}\fR will expand to the current year (without a trailing newline), and \fB#{:l:PROJNAME}\fR & \fB#{:u:PROJNAME}\fR would expand to the lowercase and uppercase versions of \fB${PROJNAME}\fR\.
.
.SS "DEFAULT FILE"
If the \fB\-d\fR option is used, skel will attempt to read a file structured like:
.
.IP "" 4
.
.nf

VAR1 the rest of the line goes in the variable
# this is a comment
VAR2 another variable
.
.fi
.
.IP "" 0
.
.P
and add its definitions to the environment before evaluating the template\.
.
.SH "BUGS"
i18n/locale handling is not supported yet\.
.
.P
Variable names are limited to 256 bytes long\.
.
.SH "COPYRIGHT"
\fBskel\fR is Copyright (C) 2012\-18 Scott Vokes \fIvokes\.s@gmail\.com\fR\.
.
.SH "SEE ALSO"
cat(1), env(1), environ(7), sh(1)


CHANGELOG.md
# skel Changes By Release

## v0.3.1 2018-07-31

### Other Improvements

Rebuilt the man pages.


## v0.3.0 2018-07-31

### API Changes

Instead of hardcoding the output buffer size, grow it on demand,
and significantly increase the size for all other buffers. 

Added optional output file argument.

If the template file argument is a directory, list the template
files in it and exit.


### Bug Fixes

When saving variables from the default file, save to the end of the
line, not just until the first whitespace.

Clear substitution attributes after expansion, rather than erroneously
applying them to all other substitutions on the same template line.

Don't modify environment variables directly -- copy their contents into
a buffer and modify that instead.


### Other Improvements

Added this CHANGELOG.

Moved code into `src/`, tests into `test/`, and build into `build/`.

Minor cleanup and code style changes.


## v0.2.0 2015-02-08

### Other Improvements

Add man page.

Internal cleanup.


## v0.1.2 2015-01-06

### API Changes

Add `-x` flag and execution patterns.

### Other Improvements

Added `install` and `uninstall` make targets.


## v0.1.1 2014-02-24

### Bug Fixes

Removed `-i` option (ignore templates in current directory) because
it combined badly with full paths.

Preserve escape '\'s that aren't before substitution markers.


## v0.1.0 2012-09-29

Initial public release.


skel.c
/* 
 * Copyright (c) 2012-14 Scott Vokes <vokes.s@gmail.com>
 *  
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *  
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <unistd.h>
#include <string.h>
#include <err.h>
#include <errno.h>
#include <limits.h>

#define SKEL_VERSION_MAJOR 0
#define SKEL_VERSION_MINOR 1
#define SKEL_VERSION_PATCH 2
/* 0.1.2 +EXEC_CHAR */

#define BUF_SZ MAX_INPUT

typedef enum { MODE_VERBATIM, MODE_IN_PATTERN } sub_mode;

#define DEF_OPEN_PATTERN "#{"
#define DEF_CLOSE_PATTERN "}"

#ifndef DEF_HOME_PATH
#define DEF_HOME_PATH "~/.dem_bones"
#endif

#ifndef DEF_SYSTEM_PATH
#define DEF_SYSTEM_PATH "/usr/local/share/skel"
#endif

/* Global settings */
static FILE *template = NULL;
static char *skel_path = NULL;
static char escape = '\\';
static char *sub_open = DEF_OPEN_PATTERN;
static char *sub_close = DEF_CLOSE_PATTERN;
static char *defaults_file = NULL;
static int abort_on_undef = 0;     /* abort on undefined substitution */
static char exec_char = '\0';      /* '\0' => off */

static void usage() {
    fprintf(stderr,
        "skel %u.%u.%u by Scott Vokes <vokes.s@gmail.com>\n"
        "usage: \n"
        "  env FOO=\"definition\" \\\n"
        "    BAR=\"other definition\" \\\n"
        "    skel [-h] [-o OPENER] [-c CLOSER] [-d FILE]\n"
        "         [-p PATH]  [-x EXEC] [-e] [TEMPLATE]\n"
        "\n"
        " -h:        help\n"
        " -o OPENER: set substitution open pattern (default \"" DEF_OPEN_PATTERN "\")\n"
        " -c CLOSER: set substitution close pattern (default \"" DEF_CLOSE_PATTERN "\")\n"
        " -x EXEC    exec patterns starting with EXEC char and insert result.\n"
        "            If doubled, trailing newlines will be stripped.\n"
        "            Example: -x %% '#{%%%%date +%%Y}' => '2015'.\n"
        " -d FILE:   set defaults file (a file w/ a list of \"KEY rest_of_line\" pairs)\n"
        " -p PATH:   path to skeleton files (closet)\n"
        " -e:        treat undefined variable as an error\n",
        SKEL_VERSION_MAJOR, SKEL_VERSION_MINOR, SKEL_VERSION_PATCH);
    exit(1);
}

static void print_env_var(char *varname) {
    char *var = NULL;
    var = getenv(varname);
    if (var == NULL) {
        if (abort_on_undef) {
            fprintf(stderr, "Undefined variable: '%s'\n", varname);
            exit(2);
        } else {
            printf("%s%s%s", sub_open, varname, sub_close);
        }
    } else {
        printf("%s", var);
    }
}

static void execute_pattern(char *cmd) {
    bool strip_nl = false;
    if (cmd[0] == exec_char) {
        strip_nl = true;
        cmd++;
    }
    while (cmd[0] == ' ') { cmd++; }  /* eat leading whitespace */

    FILE *child = popen(cmd, "r");
    if (child == NULL) {
        fprintf(stderr, "popen failure for command '%s'\n", cmd);
    } else {
        char buf[BUF_SZ];
        for (;;) {
            char *line = fgets(buf, sizeof(buf), child);
            if (line) {
                int len = strlen(line);
                if (strip_nl && len > 0 && line[len - 1] == '\n') {
                    line[len - 1] = '\0';
                }
                printf("%s", line);
            } else {
                break;
            }
        }
        pclose(child);
    }
}

static void sub_line(const char *line) {
    static char buf[BUF_SZ], var_buf[BUF_SZ];
    int i = 0, sub_i = 0, buf_i = 0, var_i = 0;
    sub_mode mode = MODE_VERBATIM;
    char c = '\0';
    
    while ((c = line[i])) {
        if (mode == MODE_VERBATIM) {
            /* This could be factored better... */
            if (c == escape && line[i + 1] == sub_open[0]) {
                buf[buf_i++] = line[++i];
            } else if (c == sub_open[0]) {     /* check for 'sub_open' match */
                for (sub_i = 0; sub_open[sub_i] == line[i + sub_i]; sub_i++) {
                    ;
                }

                if (sub_open[sub_i] == '\0') { /* reached end of pattern */
                    i += sub_i - 1;
                    mode = MODE_IN_PATTERN;
                    buf[buf_i] = '\0';
                    printf("%s", buf);
                    buf_i = sub_i = 0;
                } else {                       /* not a match */
                    memcpy(buf + buf_i, line + i, sub_i);
                    i += sub_i - 1;
                    buf_i += sub_i;
                    sub_i = 0;
                }
            } else {
                buf[buf_i++] = c;
                if (buf_i == BUF_SZ - 1) {     /* full; flush. */
                    buf[buf_i] = '\0';
                    printf("%s", buf);
                    buf_i = 0;
                }
                if (c == '\n') break;   /* c == '\0' is broken by the while */
            }
        } else if (mode == MODE_IN_PATTERN) {
            if (c == sub_close[0]) { /* check for 'sub_close' match */
                for (sub_i = 1; sub_close[sub_i] == line[i + sub_i]; sub_i++) {
                    ;
                }
                if (sub_close[sub_i] == '\0') { /* matched */
                    i += sub_i - 1;
                    var_buf[var_i] = '\0';
                    if (var_i > 1 && exec_char == var_buf[0]) {
                        execute_pattern(&var_buf[1]);
                    } else {
                        print_env_var(var_buf);
                    }
                    var_i = sub_i = 0;
                    mode = MODE_VERBATIM;
                } else {                       /* match fail */
                    memcpy(var_buf + var_i, line + i, sub_i + 1);
                    i += sub_i + 1;
                    buf_i += sub_i + 1;
                    sub_i = 0;
                }
            } else {
                var_buf[var_i++] = c;
            }
        }
        i++;
    }

    if (mode == MODE_IN_PATTERN) {
        fprintf(stderr, "End of file in pattern.\n");
        exit(1);
    }

    /* Print remaining. */
    if (buf_i > 0) { buf[buf_i] = '\0'; printf("%s", buf); }
}

/* Read the template from the file stream, line by line,
 * and print it (with variable substitutions). */
static void sub_template() {
    static char buf[BUF_SZ];
    char *pbuf = NULL;
    while ((pbuf = fgets(buf, BUF_SZ, template))) {
        sub_line(pbuf);
    }
}

static void handle_args(int *argc, char ***argv) {
    int f = 0;
    while ((f = getopt(*argc, *argv, "ho:c:d:p:ex:")) != -1) {
        switch (f) {
        case 'h':               /* help */
            usage();
        case 'o':               /* set substitution opener */
            sub_open = optarg;
            break;
        case 'c':               /* set substitution closer */
            sub_close = optarg;
            break;
        case 'd':               /* load defaults file */
            defaults_file = optarg;
            break;
        case 'p':               /* path to closet */
            skel_path = optarg;
            break;
        case 'e':               /* error on NULL substitution */
            abort_on_undef = 1;
            break;
        case 'x':               /* execute patterns */
            exec_char = optarg[0];
            break;
        case '?':
        default:
            usage();
        }
    }
    *argc -= optind;
    *argv += optind;
}

/* In order, check:
 *   ${NAME}
 *   ${SKEL_CLOSET:-~/.dem_bones}/${NAME}
 *   ${SKEL_SYSTEM_CLOSET:-/usr/local/share/skel}/${NAME}
 * */
static FILE *open_skel_file(const char *name) {
    static char pathbuf[PATH_MAX];
    char *home = getenv("HOME");
    FILE *f = NULL;
    if (skel_path == NULL) skel_path = getenv("SKEL_CLOSET");
    if (skel_path == NULL) skel_path = DEF_HOME_PATH;

    f = fopen(name, "r");
    #define TRY_PATH(fmt, ...) \
        if (PATH_MAX <= snprintf(pathbuf, PATH_MAX, fmt, __VA_ARGS__)) {\
            return NULL;                                                \
        }                                                               \
        if (0) { printf("trying path '%s'...\n", pathbuf); }            \
        f = fopen(pathbuf, "r")

    /* not found => try user's closet */
    if (f == NULL) {
        if (skel_path[0] == '~' && skel_path[1] == '/') {
            TRY_PATH("%s/%s/%s", home, skel_path + 2, name);
        } else {
            TRY_PATH("%s/%s", skel_path, name);
        }
    }

    /* still not found => try global closet */
    if (f == NULL) {
        skel_path = getenv("SKEL_SYSTEM_CLOSET");
        if (skel_path == NULL) skel_path = DEF_SYSTEM_PATH;
        TRY_PATH("%s/%s", skel_path, name);
    }

    /* not found => fail */
    if (f == NULL) return NULL;

    errno = 0;                  /* suppress intemediate failures */
    return f;
    #undef TRY_PATH
}

static void read_defaults() {
    static char buf[BUF_SZ];
    char *line = NULL;
    FILE *df = fopen(defaults_file, "r");
    char *key = NULL, *value = NULL;
    if (df == NULL) err(1, "defaults file '%s'", defaults_file);
    while ((line = fgets(buf, BUF_SZ, df))) {
        if (line[0] == '#') continue;
        if ((key = strtok(line, " \t\n"))) {
            if ((value = strtok(NULL, " \t\n"))) {
                if (setenv(key, value, 0) < 0) err(1, "setenv '%s'", line);
            }
        }
    }
}

int main(int argc, char **argv) {
    defaults_file = getenv("SKEL_DEFAULTS");
    handle_args(&argc, &argv);

    if (argc == 0 || (argc >= 1 && 0 == strcmp("-", argv[0]))) {
        template = stdin;
    } else {
        template = open_skel_file(argv[0]);
        if (NULL == template) { err(1, "skeleton file '%s'", argv[0]); }
    }

    if (defaults_file) { read_defaults(); }
    sub_template();
    return 0;
}
